Source/OxyPlot/Series/ContourSeries.cs

```diff
@@ -486,100 +459,118 @@ private void AddContourLabels(Contour contour, ScreenPoint[] pts, ICollection<Co
             contourLabels.Add(new ContourLabel { Position = pos, Angle = angle, Text = text });
         }
 
-        /// <summary>
-        /// Finds the connected segment.
-        /// </summary>
-        /// <param name="point">The point.</param>
-        /// <param name="contourLevel">The contour level.</param>
-        /// <param name="eps">The distance tolerance.</param>
-        /// <param name="reverse">reverse the segment if set to <c>true</c>.</param>
-        /// <returns>The connected segment, or <c>null</c> if no segment was found.</returns>
-        private ContourSegment FindConnectedSegment(DataPoint point, double contourLevel, double eps, out bool reverse)
-        {
-            reverse = false;
-            foreach (var s in this.segments)
-            {
-                if (!AreClose(s.ContourLevel, contourLevel, eps))
-                {
-                    continue;
-                }
-
-                if (AreClose(point, s.StartPoint, eps))
-                {
-                    return s;
-                }
-
-                if (AreClose(point, s.EndPoint, eps))
-                {
-                    reverse = true;
-                    return s;
-                }
-            }
-
-            return null;
-        }
-
         /// <summary>
         /// Joins the contour segments.
         /// </summary>
         /// <param name="eps">The tolerance for segment ends to connect (squared distance).</param>
         private void JoinContourSegments(double eps = 1e-10)
         {
-            // This is a simple, slow, naïve method - should be improved:
-            // http://stackoverflow.com/questions/1436091/joining-unordered-line-segments
             this.contours = new List<Contour>();
-            var contourPoints = new List<DataPoint>();
-            int contourPointsCount = 0;
 
-            ContourSegment firstSegment = null;
-            int segmentCount = this.segments.Count;
-            while (segmentCount > 0)
+            static IEnumerable<SegmentPoint> GetPoints(ContourSegment segment)
             {
-                ContourSegment segment1 = null, segment2 = null;
+                var p1 = new SegmentPoint(segment.StartPoint);
+                var p2 = new SegmentPoint(segment.EndPoint);
+                p1.Partner = p2;
+                p2.Partner = p1;
+                yield return p1;
+                yield return p2;
+            }
 
-                if (firstSegment != null)
+            foreach (var group in this.segments.GroupBy(p => p.ContourLevel))
+            {
+                var level = group.Key;
+                var points = group.SelectMany(GetPoints).OrderBy(p => p.Point.X).ToList();
```

Comment:
Should this be sorting by `x` and then `y`?

Reply:
In theory sorting by `y` additionally may help, but in practice I don't think it makes a difference; I wouldn't expect there to be any two identical `x` values, so there wouldn't be any ties to be resolved by additionally sorting by `y`.

Reply:
Now that I understand the while properly, this makes more sense too.